update-report: speed up automatic updates

jonchang opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This pull request disables verbose reporting of new formula versions when automatically updating before `brew install`.

Homebrew has a perceived reputation for being extremely slow, even when the cause is more likely the fault of inconsistent network connections or fundamental limitations on the speed of unpacking large tarballs to disk. A [quick search on Twitter](https://twitter.com/search?q=%22Updating%20Homebrew%22&src=typed_query&f=live) or this issue tracker (e.g. https://github.com/Homebrew/brew/issues/6490) confirms this perception.

When running `brew install wget` Homebrew will automatically run a `brew update` to ensure that users will get the newest packages, and often this is because maintainers have fixed an outstanding bug, so it's easy to see why this was implemented. However, this update step can be very slow, increasing the delay and frustration when all the user wanted to do was to install `wget` or `node` or whatever.

This pull request does two things:

1. Skips the expensive computation of new formula versions from the FormulaVersions class when running in autoupdate mode.
    - When all the user wanted to do was install `node`, they probably don't particularly care that there's a new version of `python`.
2. Disables verbose reporting of updated formulae in autoupdate mode.
    - Because we're no longer tracking new formula versions in (1), the update report will emit a list of all updated formulae, not just formulae that have new versions (i.e. are `brew upgrade`able). We can suppress this (arguably inaccurate) report while still doing all the housekeeping tasks that `brew update-report` would normally do. 

Homebrew maintainers probably run `brew update` quite often. However, users are probably not touching their Homebrew installation on a daily basis - if all you need is e.g. `node`, you might not touch Homebrew again until there's a new version of `jq` out.

![out](https://user-images.githubusercontent.com/213484/88162584-da4cb200-cc54-11ea-9f24-f41f4d98452f.png)

I ran an experiment which simulated updating Homebrew after not touching a `brew` command over X number of days. If you update Homebrew daily, the delay for updating Homebrew is usually less than 10 seconds. But past that the time to run updates quickly increases, to over a minute if you only touch Homebrew once a week. With this pull request however, this delay is around a relatively constant ~5 seconds. This was run on quite powerful hardware, and with all of the Git objects already on disk, so a slower network or mechanical disk would only increase the update delay.

MikeMcQuaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -118,7 +118,7 @@ def update_report
         puts "No changes to formulae."
       else
-        hub.dump
+        hub.dump unless args.preinstall?
```
I think this is still worth outputting just without the modified formulae? The added/deleted/renamed ones would still be of interest. Could note if needed that we're not outputting modified ones.

MikeMcQuaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -218,16 +218,18 @@ def report
-          nil
-        rescue Exception => e # rubocop:disable Lint/RescueException
-          onoe "#{e.message}\n#{e.backtrace.join "\n"}" if Homebrew::EnvConfig.developer?
+        unless Homebrew.args.preinstall?
```
```suggestion
        next if Homebrew.args.preinstall?
```

jonchang replied:
This would skip adding modified formulae to the reporter, meaning that `update-report` commands that check the reporter for changed formulae wouldn't have the right data. For example `DescriptionCacheStore` would be out of date.

jonchang commented:
This now summarizes updated formulae, rather than listing them all during auto-update.

```console
% brew update --preinstall
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> Updated Formulae
Updated 1 formula.
```

MikeMcQuaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -218,6 +223,12 @@ def report
         new_tap = tap.tap_migrations[name]
         @report[status.to_sym] << full_name unless new_tap
       when "M"
+        # Skip reporting updated formulae to speed up automatic updates.
+        if Homebrew.args.preinstall?
+          @report[:M] << tap.formula_file_to_name(src)
+          next
+        end
+
```
```suggestion
        @report[:M] << tap.formula_file_to_name(src)

```

MikeMcQuaid replied:
```suggestion
        name = tap.formula_file_to_name(src)
        
        # Skip reporting updated formulae to speed up automatic updates.
        if Homebrew.args.preinstall?
          @report[:M] << name
          next
        end

```

MikeMcQuaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -229,6 +240,7 @@ def report
         rescue Exception => e # rubocop:disable Lint/RescueException
           onoe "#{e.message}\n#{e.backtrace.join "\n"}" if Homebrew::EnvConfig.developer?
         end
+
         @report[:M] << tap.formula_file_to_name(src)
```
```suggestion
```

Bo98 replied:
I think moving this would result in the version check not having any effect.

There is some argument however that the version check stuff should be moved to where the list is printed, so that `DescriptionCacheStore` updates don't depend on version bumps.

MikeMcQuaid replied:
> I think moving this would result in the version check not having any effect.

Yeh, sorry, missed that.

MikeMcQuaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -433,6 +445,19 @@ def dump
     dump_formula_report :DC, "Deleted Casks"
   end
 
+  def dump_without_updated_formulae
```
Could this be a conditional argument on `dump(updated_formula_report: false)` instead? There's quite a bit of duplication here (and the `Key Legend` is missing).

MikeMcQuaid commented on Library/Homebrew/cmd/update-report.rb:
```diff
@@ -229,6 +240,7 @@ def report
           onoe "#{e.message}\n#{e.backtrace.join "\n"}" if Homebrew::EnvConfig.developer?
         end
+
         @report[:M] << tap.formula_file_to_name(src)
```
```suggestion
        @report[:M] << name
```

jonchang commented:
Merging, thanks for the reviews everyone
